### PR TITLE
parser, ast: add ColumnOptionCollate for all column types

### PR DESCRIPTION
**Pull request by bb7133**
### What problem does this PR solve? 
In MySQL `CREATE TABLE` syntax, column collation is part of the column option, so the following statements are valid(although meaningless):

```
create table t(a decimal collate utf8_bin)
create table t(a decimal collate utf8_bin collate utf8_general_ci)
```

This PR makes those statements acceptable

### What is changed and how it works?
Column "COLLATE" definitions are moved from 'StringType' to 'ColumnOption'

### Check List

Tests
 - Unit test

Code changes
 - Has interface methods change

Side effects
 - Increased code complexity

Related changes
 - None

### CODE REVIEW

File: parser.y [hunk 1]
```diff
@@ -1481,7 +1481,18 @@ ColumnDefList:
 ColumnDef:
 	ColumnName Type ColumnOptionListOpt
 	{
-		$$ = &ast.ColumnDef{Name: $1.(*ast.ColumnName), Tp: $2.(*types.FieldType), Options: $3.([]*ast.ColumnOption)}
+		def := &ast.ColumnDef{Name: $1.(*ast.ColumnName), Tp: $2.(*types.FieldType)}
+		opts := make([]*ast.ColumnOption, 0)
+		for _, opt := range $3.([]*ast.ColumnOption) {
```
Comment (winkyao) [inline parser.y]: Move this logic to CreateTable?

Comment (bb7133) [reply parser.y]: Sure

File: ast/ddl.go [hunk 2]
```diff
@@ -416,6 +418,12 @@ func (n *ColumnOption) Restore(ctx *RestoreCtx) error {
 		if err := n.Refer.Restore(ctx); err != nil {
 			return errors.Annotate(err, "An error occurred while splicing ColumnOption ReferenceDef")
 		}
+	case ColumnOptionCollate:
```
Comment (crazycs520) [inline ast/ddl.go]: Please add test in `TestDDL` for testing restore.

Comment (bb7133) [reply ast/ddl.go]: addressed, PTAL!

**Comment by zimulala** [comment]
Please fix the conflict.

**Comment by bb7133** [comment]
Updated, thanks @zimulala

**Pull request merged**